core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1239,25 +1258,48 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
 
     override def process(): Unit = {
       if (!isActive) return
-      val curBrokers = zkClient.getAllBrokersInCluster.toSet
-      val curBrokerIds = curBrokers.map(_.id)
+      val curBrokerAndEpochs = zkClient.getAllBrokerAndEpochsInCluster
+      val curBrokerIdAndEpochMap = curBrokerAndEpochs.map(e => (e._1.id, e._2)).toMap
+      val curBrokers = curBrokerAndEpochs.map(_._1).toSet
+      val curBrokerIds = curBrokerIdAndEpochMap.keySet
       val liveOrShuttingDownBrokerIds = controllerContext.liveOrShuttingDownBrokerIds
       val newBrokerIds = curBrokerIds -- liveOrShuttingDownBrokerIds
       val deadBrokerIds = liveOrShuttingDownBrokerIds -- curBrokerIds
+      val bouncedBrokerIds = (curBrokerIds & liveOrShuttingDownBrokerIds)
+        .filter(brokerId => curBrokerIdAndEpochMap(brokerId) > controllerContext.brokerEpochs(brokerId))
```

Comment:
I actually think of what you have suggested at the very beginning and the reason I didn't do that is that the only place we can update the broker epoch is in BrokerChange event. Broker epoch will only change when the broker ephemeral znode gets deleted and re-created so that is why we use czxid (create transaction id). czxid will only change when we create the broker znode, not when we modify it, so in `BrokerModification` event the czxid will not change. The only place we will capture the czxid change is in `BrokerChange` evnet because we are listening on the children nodes change.